warehouse/accounts/security_policy.py

```diff
@@ -117,16 +117,25 @@ def identity(self, request):
         request.add_response_callback(add_vary_callback("Cookie"))
         request.authentication_method = AuthenticationMethod.SESSION
 
-        userid = self._session_helper.authenticated_userid(request)
-        if userid is None:
-            return None
-
         # Session authentication cannot be used for uploading
         if request.matched_route.name in ["forklift.legacy.file_upload"]:
             return None
 
+        userid = self._session_helper.authenticated_userid(request)
+        if userid is None:
+            return None
+
         login_service = request.find_service(IUserService, context=None)
 
+        # A user might delete their account and immediately issue a request
+        # while the deletion is processing, causing the session check
+        # (via authenticated_userid above) to pass despite the user no longer
+        # existing. We catch that here to avoid raising during the password
+        # staleness check immediately below.
+        user = login_service.get_user(userid)
+        if user is None:
+            return None
```

Comment:
I deleted this check from the original refactor because I thought session lookup guaranteed us `user` availability; hopefully this comment (and test coverage) will prevent any future removals 🙂